File: SPA/jsonToheader.py
import glob
import os

file_list = list() 

def json2header(file_path):
    f=open(file_path, encoding='utf-8')
    content=f.read()
    f.close()
    fw = open( os.path.dirname(file_path).rstrip('/') + '/meta.h', 'w')
    header = '//A header file genereate by Sakura J2H tool\n'
    header += '//Contains the infomation of a module of Sakura Engine\n'
    header += '//With the MIT LicenseLicense Copyright!\n'
    header += '#pragma once\n'
    header += '#include <tuple>\n'
    header += '#include <cstddef>\n\n'
    header += 'const char sp_meta[] = {\n'  
    counter = 0
    for _byte in content.encode():
        if counter >= 12:
            header += '\n' 
            counter = 0 
        header += hex(_byte)
        header += ', '
        counter=counter+1
    header = header.rstrip()
    header = header.rstrip(',')
    header += '\n};\n\n'
    header += 'const std::size_t sp_meta_size = '
    header += str(len(content.encode()))
    header += ';\n\n'
    header += 'inline std::tuple<const char*, std::size_t> GetMeta(void)\n'
    header += '{\n'
    header += '    return std::tuple<const char*, std::size_t>'
    header += '(sp_meta, sp_meta_size);\n'
    header += '}\n'
    header += 'inline std::size_t GetMetaSize(void)\n{\n    return sp_meta_size;\n}\n'
    header += 'inline const char* GetMetaData(void)\n{\n    return sp_meta;\n}\n'
    fw.write(header)  
    return
def find_meta(file_path):
    if os.path.isfile(file_path):
        if file_path.find('meta.json') >= 0:
            json2header(file_path)
        else:
            print(file_path)
    else:           
        for file_ls in os.listdir(file_path):
            find_meta(os.path.join(file_path,file_ls))
    return file_list

find_meta(os.getcwd()) 

File: BuildTools/JsonToHeader.lua
lfs = require "lfs"
json = require "BuildTools/json"

JsonFileModify = nil
for file in lfs.dir(arg[1]) do
    local ff = arg[1].."/"..file
    if file == arg[2]..".json" then
        local attr = lfs.attributes(ff)
        JsonFile = ff
        JsonFileModify = attr.modification
    end
    if file == arg[2]..".h" then
        local attr = lfs.attributes(ff)
        MetaHeader = ff
        MetaHeaderModify = attr.modification
    end
end
if JsonFileModify == nil then
    print("No meta.json in the dir "..arg[1].." of Plugin.")
elseif MetaHeaderModify == nil or MetaHeaderModify <= JsonFileModify then
    print(arg[1].." Refreshing meta header.")
    JsF = io.open(JsonFile, "r")
    HdrF = io.open(arg[1].."/"..arg[2]..".h", "w")
    io.output(HdrF)
    io.write("//A header file genereate by Sakura J2H tool\n")
    io.write("//Contains the infomation of a module of Sakura Engine\n")
    io.write("//With the MIT License Copyright!\n")
    io.write("#pragma once\n")
    io.write("#include <string>\n")
    io.write("#include <cstddef>\n\n")
    io.write("const std::string sp_meta = \nR\"(")  
    io.write(JsF:read("*a")..")\";\n")
    io.write("inline const char* __GetMetaData(void)\n{\n    return sp_meta.c_str();\n}\n")
    io.write("public:\nvirtual const char* GetMetaData(void) override\n")
    io.write("{return __GetMetaData();}\n")
    io.close(JsF)
    io.close(HdrF)
end
